runtime/src/iree/hal/cts/driver_test.h

```diff
@@ -22,25 +22,99 @@ namespace cts {
 
 class driver_test : public CtsTestBase {};
 
-TEST_P(driver_test, QueryAndCreateAvailableDevices) {
+TEST_P(driver_test, QueryAndCreateAvailableDevicesByID) {
   iree_host_size_t device_info_count = 0;
   iree_hal_device_info_t* device_infos = NULL;
   IREE_ASSERT_OK(iree_hal_driver_query_available_devices(
       driver_, iree_allocator_system(), &device_info_count, &device_infos));
 
-  std::cout << "Driver has " << device_info_count << " device(s)";
+  std::cout << "Driver has " << device_info_count << " device(s)\n";
   for (iree_host_size_t i = 0; i < device_info_count; ++i) {
     std::cout << "  Creating device '"
               << std::string(device_infos[i].name.data,
                              device_infos[i].name.size)
-              << "'";
+              << "'..\n";
     iree_hal_device_t* device = NULL;
     IREE_ASSERT_OK(iree_hal_driver_create_device_by_id(
         driver_, device_infos[i].device_id, /*param_count=*/0, /*params=*/NULL,
         iree_allocator_system(), &device));
     iree_string_view_t device_id = iree_hal_device_id(device);
-    std::cout << "  Created device with id: '"
-              << std::string(device_id.data, device_id.size) << "'";
+    std::cout << "    Created device with id: '"
+              << std::string(device_id.data, device_id.size) << "'\n";
+    iree_hal_device_release(device);
+  }
+
+  iree_allocator_free(iree_allocator_system(), device_infos);
+}
+
+TEST_P(driver_test, QueryAndCreateAvailableDevicesByOrdinal) {
+  iree_host_size_t device_info_count = 0;
+  iree_hal_device_info_t* device_infos = NULL;
+  IREE_ASSERT_OK(iree_hal_driver_query_available_devices(
+      driver_, iree_allocator_system(), &device_info_count, &device_infos));
+
+  std::cout << "Driver has " << device_info_count << " device(s)\n";
+  for (iree_host_size_t i = 0; i < device_info_count; ++i) {
+    std::cout << "  Creating device '"
+              << std::string(device_infos[i].name.data,
+                             device_infos[i].name.size)
+              << "'..\n";
+    iree_hal_device_t* device = NULL;
+    IREE_ASSERT_OK(iree_hal_driver_create_device_by_ordinal(
+        driver_, i, /*param_count=*/0, /*params=*/NULL, iree_allocator_system(),
+        &device));
+    iree_string_view_t device_id = iree_hal_device_id(device);
+    std::cout << "    Created device with id: '"
+              << std::string(device_id.data, device_id.size) << "'\n";
+    iree_hal_device_release(device);
+  }
+
+  iree_allocator_free(iree_allocator_system(), device_infos);
+}
+
+TEST_P(driver_test, QueryAndCreateAvailableDevicesByPath) {
+  iree_host_size_t device_info_count = 0;
+  iree_hal_device_info_t* device_infos = NULL;
+  IREE_ASSERT_OK(iree_hal_driver_query_available_devices(
+      driver_, iree_allocator_system(), &device_info_count, &device_infos));
+
+  std::cout << "Driver has " << device_info_count << " device(s)\n";
+
+  if (device_info_count > 0) {
```

Comment:
can you unnest these blocks? `if (device_info_count == 0) GTEST_SKIP() << "no devices available";` will skip the test with a nice message and then let you remove this if.
right now it's harder to see what's happening because of this if and then the implicit if in the loop below